KIP-42 Add producer and consumer interceptors

d1egoaz opened the pull request:
This PR includes:
Producer: `onSend` but it doesn't implement `onAcknowledgement`
Consumer: `onConsume` but it doesn't implement `onCommit`

I'm not sure if I need to add the `onClose` method. Maybe in another
iteration ¯\_(ツ)_/¯

TODO: functional test, but it looks the way sarama functional test works is that they don't have this detail.

Commit:
Add docker-composed based functional test harness

* Functional tests are skipped except when the `functional` build tag
  is passed to the go test command (i.e. go test -tags=functional)
* If TOXIPROXY_ADDR is not set when the functional tests are being run,
  it will use docker-compose to automatically spin up a
  zookeeper/kafka/toxiproxy environment suitab le for running the tests.
  This requies a working Docker and for the docker-compose command line
  tool to be installed.
* If TOXIPROXY_ADDR and KAFKA_VERSION are set, then the tests will not
  spin up any docker infrastructure and will instead rely on whatever
  kafka broker is behind the specified toxiproxy.

Commit:
Map confluent platform -> kafka version

Commit:
Use the built-in docker compose support to run CI tests

Commit:
Delete vagrant test harness setup

Commit:
Run the linter across functional tests as well

Commit:
Make linter pass on functional tests

d1egoaz committed:
KIP-42 Add producer and consumer interceptors

This PR includes:
Producer: `onSend` but it doesn't implement `onAcknowledgement`
Consumer: `onConsume` but it doesn't implement `onCommit`

I'm not sure if I need to add the `onClose` method. Maybe in another
iteration ¯\_(ツ)_/¯

d1egoaz commented on interceptors.go:
```diff
@@ -0,0 +1,23 @@
+	// Please avoid modifying the message until it's safe to do so, as you'll be
+	// modifying the message reference.
+	onConsume(*ConsumerMessage)
+}
```
some comments here:
I decided to use the object reference here to avoid the need to clone hence consume more memory when dealing with each msg.
The effect of this, is that an interceptor that changes the value and for some reason fails, the value will be changed on the reference, and the next interceptor will have the value changed.

I'm not sure we'll need to strictly follow java/ruby pattern here, as with c/go you know what are the effects of changing a reference.

happy to hear your thoughts

vvuibert replied:
I think that a bad interceptor shouldn't break the chain and the next interceptor should get the same message that the bad interceptor got

psycotica0-shopify replied:
Ah, is that what you meant by "when it's safe" 😄?

Right, so mutating the message is more memory efficient, and more importantly Go doesn't have any kind of "duplicate" method built-in, because it has basically nothing built in.
So if we wanted to support this we'd need to write a duplicate message for `ConsumerMessage` and `ProducerMessage` and keep them up to date as those structures change, or else we'll be silently corrupting things.

Hmm. 🤔

psycotica0-shopify replied:
And in Ruby, any function call could throw an error and the whole stack unwinds.
That's dangerous.

In go, you need to explicitly write:
```go
if value, err := someMethod; err != nil {
  return
}
```

Where you explicitly get an error, ignore it, and return early.
I assume the author of that code would realize the things they changed before the `return` are still changed, since the code doesn't allow you to return an error, and doesn't allow you to return a new value.

Hmm. 🤔

d1egoaz replied:
yeah to all of those @psycotica0-shopify

I spent a few brain cycles on the signature of these methods.

like, why I would make the user return an error if basically I'd just ignore it and log it with the very basic logging system that Sarama has (one of the big Sarama issues, everything is logged on the Info log).

d1egoaz replied:
I'm relying to much about how Go works, that might be a little different from the KIP specification, which might not be necessarily a bad thing

vvuibert commented:
>I'm not sure if I need to add the onClose

I think we should add the onClose since we use it to close the trace exporter
https://github.com/Shopify/shopify-docker-images/blob/master/kafka-mirrormaker/tracing/src/main/java/com/shopify/kafka/tracing/TracingConsumerInterceptor.java#L62-L63

vvuibert commented on interceptors.go:
```diff
@@ -0,0 +1,23 @@
+// https://cwiki.apache.org/confluence/display/KAFKA/KIP-42%3A+Add+Producer+and+Consumer+Interceptors#KIP42:AddProducerandConsumerInterceptors-Motivation
+type ProducerInterceptor interface {
+
+	// onSend is called when the producer message is intercepted.
+	// Please avoid modifying the message until it's safe to do so, as you'll be
+	// modifying the message reference.
```
what's the message reference?

psycotica0-shopify replied:
What does "until it's safe to do so" mean?

Will that make sense to the user? I'm not sure I know when it's safe 😄

d1egoaz replied:
you got it here: https://github.com/d1egoaz/sarama/pull/1#discussion_r439616140

psycotica0-shopify commented on async_producer_test.go:
```diff
@@ -1230,6 +1231,104 @@ func TestBrokerProducerShutdown(t *testing.T) {
+			t.Error(msg.Err)
+		case msg := <-producer.Successes():
+			v, _ := msg.Value.Encode()
+			expected := TestMessage + strconv.Itoa(i) + strconv.Itoa(j)
```
You _could_ do:
```go
for i := 0; i < 10; i = i+1 {
  [...]
  expected := TestMessage + strconv.Itoa(i) + strconv.Itoa(i + 1000)
  [...]
}
```

If you wanted

d1egoaz replied:
ohh yeah right, Initially the code did a different operation and it wasn't just incrementing the index, now it does and your code makes more sense

psycotica0-shopify commented on config.go:
```diff
@@ -229,6 +229,15 @@ type Config struct {
+		// cluster.
+		// *ProducerMessage returned from the first interceptor's
+		// onSend() is passed to the second interceptor onSend(), and so on in
+		// the interceptor chain.
```
The docs here say "returned", but the interceptors don't return anything.

psycotica0-shopify commented on consumer_test.go:
```diff
@@ -1342,3 +1343,108 @@ func Test_partitionConsumer_parseResponse(t *testing.T) {
+func (b *appendInterceptor) onConsume(msg *ConsumerMessage) {
+	msg.Value = []byte(string(msg.Value) + strconv.Itoa(b.i))
+	b.i++
+}
```
Should these all be defined together somewhere? It seems weird that over in this file we're adding methods to the type we assume is defined in another file?

Maybe that's a normal Go thing, though. If so, carry on 😄

d1egoaz replied:
not necessarily, I had the same issue when navigating go source, then I read that the tooling can give you all the methods that a type implements, and you shouldn't rely only in what's defined in a file.
but if that helps reading the code, I'll move it.

this is what the tooling currently shows:
```
/Users/diegoalvarez/src/github.com/Shopify/sarama/async_producer_test.go:1234.6-1234.22: definition of type appendInterceptor (size 8, align 8)
/Users/diegoalvarez/src/github.com/Shopify/sarama/async_producer_test.go:1234.6-1234.22: Methods:
/Users/diegoalvarez/src/github.com/Shopify/sarama/consumer_test.go:1347.29-1347.37: 	method (*appendInterceptor) onConsume(msg *ConsumerMessage)
/Users/diegoalvarez/src/github.com/Shopify/sarama/async_producer_test.go:1238.29-1238.34: 	method (*appendInterceptor) onSend(msg *ProducerMessage)
/Users/diegoalvarez/src/github.com/Shopify/sarama/async_producer_test.go:1234.6-1234.22: Fields:
/Users/diegoalvarez/src/github.com/Shopify/sarama/async_producer_test.go:1235.2-1235.2: 	i int
```

d1egoaz commented:
> think we should add the onClose since we use it to close the trace exporter
> https://github.com/Shopify/shopify-docker-images/blob/master/kafka-mirrormaker/tracing/src/main/java/com/shopify/kafka/tracing/TracingConsumerInterceptor.java#L62-L63

still unsure about this.
I understand that most of the java code needs something like this to allow this kind of plugins to be closed successfully as this might be things that are loaded dynamically, injected, etc.
In go there is not such thing, and you can always do some kind of `defer myInterceptor.Close` on the method that initializes the interceptor to close things if needed.

d1egoaz committed:
include zstd on the functional tests

d1egoaz commented on functional_producer_test.go:
```diff
@@ -25,6 +26,12 @@ func TestFuncProducingGzip(t *testing.T) {
+	config.Producer.Compression = CompressionZSTD
+	testProducingMessages(t, config)
+}
+
```
https://github.com/Shopify/sarama/pull/1726

vvuibert commented on consumer_test.go:
```diff
@@ -1342,3 +1343,112 @@ func Test_partitionConsumer_parseResponse(t *testing.T) {
+			t.Error(err)
+		}
+	}
+
```
could you also close the `interceptors`

d1egoaz replied:
there is nothing to close 🤷

d1egoaz committed:
zstd needs client >= 2.1.0.0

d1egoaz committed:
Merge pull request #1726 from Shopify/diego_zstd-functional-test

Include zstd on the functional tests

vvuibert commented on interceptors.go:
```diff
@@ -0,0 +1,43 @@
+// ProducerInterceptor allows you to intercept (and possibly mutate) the records
+// received by the producer before they are published to the Kafka cluster.
+// https://cwiki.apache.org/confluence/display/KAFKA/KIP-42%3A+Add+Producer+and+Consumer+Interceptors#KIP42:AddProducerandConsumerInterceptors-Motivation
+type ProducerInterceptor interface {
```
do we need a `ProducerInterceptor` and `ConsumerInterceptor`? could we just have an `Interceptor` with an `onCall(message)`?

d1egoaz replied:
we would need at least 2 methods, one that receives a `ProducerMessage` and other for `ConsumerMessage`, and that might be an issue, as if someone would like to only intercept a producer, it'll need to implement the `onSend` for consumer too.

So I think it's safe to leave it in 2 separated interfaces

vvuibert commented on functional_producer_test.go:
```diff
@@ -25,6 +26,12 @@ func TestFuncProducingGzip(t *testing.T) {
 	testProducingMessages(t, config)
 }
 
+func TestFuncProducingZstd(t *testing.T) {
```
maybe rebase since https://github.com/Shopify/sarama/pull/1726 was merged

vvuibert commented on functional_producer_test.go:
```diff
@@ -174,6 +181,78 @@ func TestFuncProducingIdempotentWithBrokerFailure(t *testing.T) {
+		case err := <-consumer.Errors():
+			t.Error(err)
+		case msg := <-consumer.Messages():
+			// producer interceptors: strconv.Itoa(i) + strconv.Itoa(i+100)
+			// consumer interceptor: strconv.Itoa(i+20)
```
are these comments needed?

d1egoaz replied:
good point, I added them to make sure the reader understand the expectation in the following line, but just realized that I can introduce some variables and get rid of the comment.

changed to:
```go
			prodInteExpectation := strconv.Itoa(i) + strconv.Itoa(i+100)
			consInteExpectation := strconv.Itoa(i + 20)
			expected := TestMessage + prodInteExpectation + consInteExpectation
```

vvuibert commented on interceptors.go:
```diff
@@ -0,0 +1,43 @@
+// https://cwiki.apache.org/confluence/display/KAFKA/KIP-42%3A+Add+Producer+and+Consumer+Interceptors#KIP42:AddProducerandConsumerInterceptors-Motivation
+type ConsumerInterceptor interface {
+
+	// onConsume is called when the consumed message is intercepted. Please
+	// avoid modifying the message until it's safe to do so, as this is _not_ a
+	// copy of the message.
```
this is a bit scary since monitoring should not affect the code. this doc implies that an error in monitoring could break the code

d1egoaz replied:
yeah, that was my opinionated take on this.
I added some comments on the upstream PR about this, let's see if we can get some 👀 on this.

d1egoaz replied:
> this doc implies that an error in monitoring could break the code

Not break the code, but modify partially the message

d1egoaz committed:
make linter happy

d1egoaz committed:
remove code comments

d1egoaz committed:
use the third person singular present tense

s/have/has

Commit:
Remove redundant hard-coding of env.KafkaVersion

bai committed:
Merge pull request #1695 from zendesk/ktsanaktsidis/docker_compose_tests

Use docker-compose to run the functional tests

wclaeys committed:
Add support for manual commit to ConsumerGroup

- expose a `Commit()` sync method on ConsumerGroupSession
- don't create mainLoop in OffsetManager unless AutoCommit is enabled

dnwe committed:
Merge pull request #1699 from wclaeys/master

Consumer group support for manually comitting offsets

d1egoaz committed:
export interceptor methods

d1egoaz committed:
fix renamed variable after merging latest master
